### PR TITLE
Fixed #30686 -- Used Python HTMLParser in utils.text.Truncator

### PR DESCRIPTION
**Pull request by smithdc1**
Marking as draft -- it's not finished but thought it may be helpful to share my thinking as the ticket is being reviewed. Even if the proposal isn't valid, maybe the extra tests are worth keeping. 

https://code.djangoproject.com/ticket/30686

### CODE REVIEW

File: tests/utils_tests/test_text.py [hunk 1]
```diff
@@ -95,17 +95,72 @@ def test_truncate_chars(self):
...
+        self.assertEqual(
+            "<br>The <hr/>quick <em>brown…</em>", truncator.chars(16, html=True)
+        )
+
+        # Test html entities
+        truncator = text.Truncator(
+            "<i>Buenos d&iacute;as! &#x00bf;C&oacute;mo est&aacute;?</i>"
+        )
+        self.assertEqual(
+            "<i>Buenos días! ¿Cómo está?</i>",
```
Comment (smithdc1) [inline tests/utils_tests/test_text.py:151]: So currently we get `í` in place of `&iacute;`. 

It's possible to convert back to HTML elements (?) with `encode('ascii', 'xmlcharrefreplace')` but that gives me back `&#237;`. 🤔 Likely need to decide if this approach is even worthwhile before trying to answer this.

File: django/utils/text.py [hunk 2]
```diff
@@ -62,6 +64,68 @@ def _generator():
...
+    def handle_endtag(self, tag):
+        if not self.complete and tag not in self.void_elements:
+            self.output += f"</{tag}>"
+            self.tags.remove(tag)
+
+    def handle_data(self, data):
+        if self.complete:
+            return
+        if self.func == "words":
+            # TODO swap this round to find Match Groups?
```
Comment (smithdc1) [inline django/utils/text.py:105]: I was wondering if with a different regex we can combine the two lines which look something like this `output = escape(" ".join(data[: self.truncate_len]))` but have a different string as the base of the `join()`

File: tests/utils_tests/test_text.py [hunk 3]
```diff
@@ -95,17 +95,72 @@ def test_truncate_chars(self):
         )
 
     def test_truncate_chars_html(self):
-        perf_test_values = [
```
Comment (smithdc1) [inline tests/utils_tests/test_text.py:98]: Some of these are not valid HTML so don't get through the HTML parser. e.g 50,000 `&`s. I'd need a bit of help to understand what exactly this is testing?

File: tests/utils_tests/test_text.py (hunk 1 above)
Comment (felixxm) [inline tests/utils_tests/test_text.py:151]: :wink: 
```suggestion
            "<i>Buenos días! ¿Cómo estás?</i>",
```

**Comment by smithdc1** [comment]
We should bear in mind 7f65974f8219729c047fbbf8cd5cc9d80faefe77

File: tests/utils_tests/test_text.py (hunk 1 above)
Comment (matthiask) [reply tests/utils_tests/test_text.py:151]: I think in the context of truncating í is more useful than `&iacute;` since the client will only show one char and also because truncating inside an entity is made impossible by converting entities to characters.

I wouldn't advocate for removing entities everywhere but in the context of displaying a truncated result to users it should be fine. 

(Converting back to entities came up a few times; a person who wanted it found a partial solution here using BeautifulSoup https://github.com/matthiask/html-sanitizer/issues/23 )

**Comment by smithdc1** [comment]
Having checked out my branch / PR to `django-asv` (https://github.com/django/django-asv/pull/70) and running `asv continuous main 9af6c3b20a55d62e5c3a8b4425a4593a915332a6 -b utils_benchmarks` I get:

```
       before           after         ratio
     [e1a093f8]       [9af6c3b2]
     <main>           <ticket_30686_2>
-     2.20±0.01ms         1.68±0ms     0.76  utils_benchmarks.truncator.benchmark.TruncatorBenchmark.time_chars_long
-      34.2±0.2ms       24.5±0.2ms     0.71  utils_benchmarks.truncator.benchmark.TruncatorBenchmark.time_chars_long_html
-        1.63±0ms      1.10±0.01ms     0.68  utils_benchmarks.truncator.benchmark.TruncatorBenchmark.time_chars_short
-      78.9±0.8ms         52.8±1ms     0.67  utils_benchmarks.truncator.benchmark.TruncatorBenchmark.time_words_long_html
-      22.4±0.3ms      9.90±0.08ms     0.44  utils_benchmarks.truncator.benchmark.TruncatorBenchmark.time_words_short_html

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

**Comment by carltongibson** [comment]
@matthiask you're the domain expert here. Can I ask you to (re-)review with a mind to _are we doing it?_ — thanks! 🎁

**Comment by ngnpope** [comment]
I added some suggested changes in https://github.com/smithdc1/django/pull/3.

File: tests/utils_tests/test_text.py (hunk 1 above)
Comment (akshaykedar1996) [reply tests/utils_tests/test_text.py]: I'm glad to provide you with more information on the change made in Django's codebase!

In Django, the Truncator class in the django.utils.text module is used to truncate text. Prior to the change made in ticket #16421, Truncator used regular expressions to parse and truncate HTML content. However, regular expressions are not a reliable way to parse HTML, and may lead to unexpected results or errors.

As a result, in ticket #30686, the Truncator class was updated to use Python's built-in HTMLParser module instead of regular expressions. This change made the Truncator class more reliable and accurate when handling HTML content.

Using the HTMLParser module, Truncator can now properly handle nested HTML tags and attributes, and avoid accidentally cutting off tags or attributes in the middle. The HTMLParser module is a more robust and flexible way to handle HTML parsing, making Truncator more reliable and accurate when handling HTML content.

Overall, this change improved the accuracy and reliability of the Truncator class when handling HTML content, which is an important feature for many Django applications.

**Comment by carltongibson** [comment]
> Some of these are not valid HTML so don't get through the HTML parser. e.g 50,000 &s. I'd need a bit of help to understand what exactly this is testing? (https://github.com/django/django/pull/16421#discussion_r1060961951)

> We should bear in mind https://github.com/django/django/commit/7f65974f8219729c047fbbf8cd5cc9d80faefe77

OK... so... what motivated the whole ticket here was a persistent trickle of security reports about ever more ingenious ways of making the regex misbehave. That problem is intractable, hence _Let's use a parser_. As far as I can see, adjusting the tests there to show that the input is (quickly) rejected would be sufficient. (We need to ensure that we don't open a DoS vector, not necessarily handle such deviant input.) //cc @django/django-security-team

**Comment by carltongibson** [comment]
https://github.com/django/django/pull/16421#issuecomment-1409894531 — Using the parser is quicker. I think we're green. 

@smithdc1 if you want to look at @ngnpope's comments, and resolve, we should be able to push this forwards. 

Thanks for effort! 🏅 

(I will mark as `PNI` on the ticket for the moment.)

**Comment by matthiask** [comment]
Sorry for not reacting sooner here. The speed up is a pleasant and unexpected surprise. I do not have any further comments apart from LGTM, thanks!

**Comment by claudep** [comment]
Looks great! May I ask about how invalid html is treated before and after the patch?

File: django/utils/text.py [hunk 4]
```diff
@@ -64,6 +67,76 @@ def _generator():
...
+        "embed",
+        "hr",
+        "img",
+        "input",
+        "link",
+        "meta",
+        "source",
+        "track",
+        "wbr",
+    }
```
Comment (felixxm) [inline django/utils/text.py:88]: We have a subclass of `HTMLParser` in `django.test`:

https://github.com/django/django/blob/32d4b61c313be5169137047e9fb3668da20a5d89/django/test/html.py#L202-L222

Can we reuse it? or create a class/mixin they can share?

File: django/utils/text.py [hunk 5]
```diff
@@ -64,6 +67,76 @@ def _generator():
...
+        if func not in ("words", "chars"):
+            raise AttributeError(f"func must be 'words' or 'chars', got '{func}'.")
+        if num <= 0:
+            if func == "words":
+                return ""
+            else:
+                return truncate
+        self.func = func
+        self.num = num
+        self.truncate = truncate
```
Comment (adamchainz) [inline django/utils/text.py:132]: can you move all the arguments up into `__init__` ? it's hard to read the code to see where `self.func` etc. come from at the moment, on first reading I thought they were comign from `HTMLParser`.

File: django/utils/text.py [hunk 6]
```diff
@@ -64,6 +67,76 @@ def _generator():
...
+
+    def handle_starttag(self, tag, attrs):
+        self.output += self.get_starttag_text()
+        if tag not in self.void_elements:
+            self.tags.appendleft(tag)
+
+    def handle_endtag(self, tag):
+        if tag not in self.void_elements:
+            self.output += f"</{tag}>"
+            self.tags.remove(tag)
```
Comment (adamchainz) [inline django/utils/text.py:103]: This will raise `ValueError` on input `foo</p>` - which is technically valid HTML for a paragraph... https://htmlparser.info/parser/#implied-tags

Perhaps we should `try/.../except ValuError/pass` ?

File: django/utils/text.py (hunk 5 above)
Comment (ngnpope) [reply django/utils/text.py:132]: See my pull request to this branch posted above. Already handles this.

> I added some suggested changes in [smithdc1#3](https://github.com/smithdc1/django/pull/3).

File: django/utils/text.py (hunk 6 above)
Comment (smithdc1) [reply django/utils/text.py:131]: TIL! Although I'm not sure I completely follow this as elsewhere it says that the an opening `<p>` is not omissible? https://htmlparser.info/syntax/#optional-tags

Nevertheless it is possible that closing tags are provided without an opening tag and so we should try and do something with that rather than just raising `ValueError`.

File: tests/template_tests/filter_tests/test_truncatewords_html.py [hunk 7]
```diff
@@ -24,15 +24,15 @@ def test_truncate2(self):
...
+            '<p>one <a href="#">two - three <br> …</a></p>',
         )
 
     def test_truncate3(self):
         self.assertEqual(
             truncatewords_html(
                 '<p>one <a href="#">two - three <br>four</a> five</p>', 5
             ),
-            '<p>one <a href="#">two - three <br>four …</a></p>',
+            '<p>one <a href="#">two - three <br>four</a> …</p>',
```
Comment (ngnpope) [inline tests/template_tests/filter_tests/test_truncatewords_html.py:35]: This is different, but I think a change for the better. Not sure whether we need a backward incompatibility note just to warn that there may be small differences, but that, in general, the output is better and that the implementation is more robust?

File: tests/utils_tests/test_text.py [hunk 8]
```diff
@@ -95,17 +95,97 @@ def test_truncate_chars(self):
...
+            truncator.chars(10, html=True),
+        )
+        self.assertEqual(
+            "…",
+            truncator.chars(1, html=True),
+        )
+        self.assertEqual(
+            "…",
+            truncator.chars(0, html=True),
+        )
```
Comment (ngnpope) [inline tests/utils_tests/test_text.py:128]: This seems wrong and unexpected. If we're requesting a length of zero, we should have an empty string.

File: django/utils/text.py [hunk 9]
```diff
@@ -64,6 +88,94 @@ def _generator():
...
+
+
+class TruncateCharsHTMLParser(TruncateHTMLParser):
+    def process(self, data):
+        output = escape("".join(data[: self.remaining]))
+        return data, output
+
+    def feed(self, data):
+        if self.remaining <= 0:
+            self.output += self.truncate
```
Comment (ngnpope) [inline django/utils/text.py:162]: Following on from the above comment about zero length should give an empty string, this is the bit that causes that. Further, this seems strange. It's also potentially broken as `self.truncate` could, in theory, contain the `%(truncated_text)s` placeholder -- see the `add_truncation_text()` function.

I suggest that we remove both `TruncateCharsHTMLParser.feed()` and `TruncateWordsHTMLParser.feed()` and add the following to the top of the parent `TruncateHTMLParser.feed()`:

```python
        if self.remaining <= 0:
            return
```

File: django/utils/text.py [hunk 10]
```diff
@@ -101,13 +198,17 @@ def chars(self, num, truncate=None, html=False):
 
         # Calculate the length to truncate to (max length - end_text length)
         truncate_len = length
-        for char in self.add_truncation_text("", truncate):
+        truncate = add_truncation_text("", truncate)
+        for char in truncate:
```
Comment (ngnpope) [inline django/utils/text.py:202]: I think that this change ought to be reverted. It prematurely resolves the `%(truncated_text)s` placeholder by overriding the `truncate` output. This was probably again related to the zero length issue above.

File: tests/template_tests/filter_tests/test_truncatechars_html.py [hunk 11]
```diff
@@ -41,7 +41,7 @@ def test_truncate_unicode(self):
         )
 
     def test_truncate_something(self):
-        self.assertEqual(truncatechars_html("a<b>b</b>c", 3), "a<b>b</b>c")
+        self.assertEqual(truncatechars_html("a<b>b</b>c", 3), "a<b>b</b>…")
```
Comment (ngnpope) [inline tests/template_tests/filter_tests/test_truncatechars_html.py:44]: Hmm... This looks wrong. We should return the full output if the length is exact. Maybe we've got an off-by-one issue somewhere?

File: tests/template_tests/filter_tests/test_truncatewords_html.py [hunk 12]
```diff
@@ -24,15 +24,15 @@ def test_truncate2(self):
             truncatewords_html(
                 '<p>one <a href="#">two - three <br>four</a> five</p>', 4
             ),
-            '<p>one <a href="#">two - three …</a></p>',
+            '<p>one <a href="#">two - three <br> …</a></p>',
```
Comment (ngnpope) [inline tests/template_tests/filter_tests/test_truncatewords_html.py:27]: Perhaps unfortunate, especially given it's a `<br>`, but likely more strictly correct. Otherwise, not sure if we can avoid trailing void elements just before the replacement?

File: tests/template_tests/filter_tests/test_truncatewords_html.py [hunk 13]
```diff
@@ -53,7 +53,7 @@ def test_truncate_complex(self):
             truncatewords_html(
                 "<i>Buenos d&iacute;as! &#x00bf;C&oacute;mo est&aacute;?</i>", 3
             ),
-            "<i>Buenos d&iacute;as! &#x00bf;C&oacute;mo …</i>",
+            "<i>Buenos días! ¿Cómo …</i>",
```
Comment (ngnpope) [inline tests/template_tests/filter_tests/test_truncatewords_html.py:56]: Again, probably worth expanding the backward compatibility note I recommended above to explicitly mention the expansion of HTML entities. We've already agreed this is an acceptable trade-off and, in general, there should be little issues as long as everyone is using UTF-8 as they very much should be! _(If anyone is relying on anything else they'll have to write a custom filter to convert things back to entities as required. So be it.)_

File: tests/utils_tests/test_text.py (hunk 3 above)
Comment (ngnpope) [reply tests/utils_tests/test_text.py:98]: If it's not valid, I wouldn't worry. These were likely just stress testing the old parser to avoid ReDOS issues.

File: tests/utils_tests/test_text.py [hunk 14]
```diff
@@ -95,17 +95,97 @@ def test_truncate_chars(self):
...
+        self.assertEqual(
+            '<p>The quick <a href="xyz.html"\n id="mylink">brow…</a></p>',
+            truncator.chars(15, html=True),
+        )
+        self.assertEqual(
+            "<p>Th…</p>",
+            truncator.chars(3, html=True),
+        )
+
+        # Test self-closing tags
```
Comment (ngnpope) [inline tests/utils_tests/test_text.py:152]:
```suggestion
        # Test void elements
```

Comment (ngnpope) [reply tests/utils_tests/test_text.py:152]: Or better yet, split into a separate `test_truncate_chars_html_with_void_elements()` function.

File: tests/utils_tests/test_text.py [hunk 15]
```diff
@@ -95,17 +95,97 @@ def test_truncate_chars(self):
...
+        )
+        self.assertEqual(
+            "<br>The <hr/>quick <em>brown…</em>", truncator.chars(16, html=True)
+        )
+        self.assertEqual("<br>The <hr/>q…", truncator.chars(6, html=True))
+        self.assertEqual("<br>The <hr/>…", truncator.chars(5, html=True))
+        self.assertEqual("<br>The…", truncator.chars(4, html=True))
+        self.assertEqual("<br>Th…", truncator.chars(3, html=True))
+
+        # Test html entities
```
Comment (ngnpope) [inline tests/utils_tests/test_text.py:168]: Perhaps split into a separate `test_truncate_chars_html_with_html_entities()` function.

File: tests/utils_tests/test_text.py [hunk 16]
```diff
@@ -152,6 +236,10 @@ def test_truncate_html_words(self):
             '<p>The quick <a href="xyz.html"\n id="mylink">brown…</a></p>',
             truncator.words(3, html=True),
         )
+        self.assertEqual(
+            "<p>The…</p>",
+            truncator.words(1, html=True),
+        )
```
Comment (ngnpope) [inline tests/utils_tests/test_text.py:242]: All of these new tests and assertions should be added up front as a preliminary commit so that we can see what adjustments are made in the main commit when migrating to the new parser.

**Review by ngnpope — CHANGES_REQUESTED** [review]
Hi @smithdc1! Thanks for the update. I've had another thorough pass through this. Let me know if you want any help getting this progressed 🙂

File: django/utils/text.py [hunk 17]
```diff
@@ -12,6 +15,27 @@
...
+    "link",
+    "meta",
+    "param",
+    "source",
+    "track",
+    "wbr",
+    # Deprecated tags
+    "frame",
+    "spacer",
+}
```
[the remark this answers is not in the document]
Comment (smithdc1) [reply django/utils/text.py:37]: I'm struggling to move this. With `VOID_ELEMENTS` in `django.utils.html` the tests crash with:

```
  File "C:\Users\smith\projects\django\django\utils\text.py", line 14, in <module>
    from django.utils.html import VOID_ELEMENTS
ImportError: cannot import name 'VOID_ELEMENTS' from partially initialized module 'django.utils.html' (most likely due to a circular import) (C:\Users\smith\projects\django\django\utils\html.py)
```

Any ideas?

File: tests/utils_tests/test_text.py (hunk 16 above)
Comment (smithdc1) [reply tests/utils_tests/test_text.py:242]: I _think_ I've already done this as these tests are in this commit here 🤔  https://github.com/django/django/pull/16421/commits/fdc15362d6706ff6cef7927b6b6617a9cce8a1b8 

I'll also work on adding Adam's `"foo</p>"` test to this up front commit as well.

File: tests/utils_tests/test_text.py (hunk 8 above)
Comment (smithdc1) [reply tests/utils_tests/test_text.py:128]: I don't disagree! However, this is part of the commit which is adding tests to the current behavior. These differences made working on this more complicated!

If we're happy with a breaking change then we can simplify as you suggested.

File: django/utils/text.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (ngnpope) [reply django/utils/text.py:37]: Hmm. Annoying that there's a circular import somewhere, especially as that's the logical place to put this. _(Will have to have a look, but not at laptop right now.)_

File: tests/utils_tests/test_text.py (hunk 8 above)
Comment (ngnpope) [reply tests/utils_tests/test_text.py:128]: So I think we add the tests for the current behaviour and then intentionally break it in the main commit to return empty string which makes more sense.

File: tests/utils_tests/test_text.py (hunk 16 above)
Comment (ngnpope) [reply tests/utils_tests/test_text.py:242]: Sorry, yes, noticed this after the review 👍🏻

File: django/utils/text.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (ngnpope) [reply django/utils/text.py:37]: Ah, ok. So it looks like there is an import of `django.utils.text.normalize_newlines()` in `django.utils.html` and we're wanting to import `django.utils.html.VOID_ELEMENTS` in `django.utils.text`. I suspect that we'll need to push down one of these imports...

**Option 1:** Move the `from django.utils.text import normalize_newlines` import into `django.utils.html.linebreaks()`.

**Option 2:** Add the following to `TruncateHTMLParser` to avoid the circular import (and re-importing multiple times in different methods):

```python
class TruncateHTMLParser(HTMLParser):
    ...

    @cached_property
    def void_elements(self):
        from django.utils.html import VOID_ELEMENTS

        return VOID_ELEMENTS

    ...
```

[the remark this answers is not in the document]
Comment (smithdc1) [reply django/utils/text.py:37]: Looking at moving and `VOID_ELEMENTS` again this morning. I think we should add a backward compatibility note. I know it's not documented, but someone is going to be using it!

I also noticed that this will mean `VOID_ELEMENTS` is in `django.utils.html` but `BOOLEAN_ATTRIBUTES` (which also seems undocumented to me) would be in `django.test.html`. That feels a little inconsistent to me. Maybe we should move both, but that makes me even more nervous re backward compatibility.

File: django/utils/text.py (hunk 9 above)
Comment (smithdc1) [reply django/utils/text.py]: I tried to fix the zero length case in https://github.com/django/django/pull/16421/commits/c4f609517b99c5b39b934de0c94c7f1ae0f61b43

What do you think?

I needed to manage both the case when the length is actualy zero and when the length of `replacement` is the equal to or greater than `length`.

File: tests/template_tests/filter_tests/test_truncatechars_html.py (hunk 11 above)
Comment (smithdc1) [reply tests/template_tests/filter_tests/test_truncatechars_html.py:44]: I'm not quite sure of the correct wording to use here. 

I think this is some kind of look ahead issue where we have reached the required length and truncate it. At this point there may/may not be more text remaining in the data. Maybe we could look at the parser's `rawdata` and parse that to see if there's anything left. I'm not sure that additional complexity is worth it. We've already added a note to say there will be small changes in the output.

File: tests/utils_tests/test_text.py (hunk 8 above)
Comment (felixxm) [reply tests/utils_tests/test_text.py]: I'd still move it to the last commit. We're not changing here a behavior that can be expected by anyone, this is a regression test that should be a part of the bugfix.

**Comment by felixxm** [comment]
Please rebase.

File: django/utils/text.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (felixxm) [reply django/utils/text.py]: I'd leave `BOOLEAN_ATTRIBUTES` in `django.test.html`, it's not worth moving.

File: django/utils/text.py [hunk 18]
```diff
@@ -137,8 +218,13 @@ def words(self, num, truncate=None, html=False):
         """
         self._setup()
         length = int(num)
+        if length == 0:
+            return ""
         if html:
-            return self._truncate_html(length, truncate, self._wrapped, length, True)
+            parser = TruncateWordsHTMLParser(length=num, replacement=truncate)
```
Comment (ngnpope) [inline django/utils/text.py:224]:
```suggestion
            parser = TruncateWordsHTMLParser(length=length, replacement=truncate)
```

File: tests/template_tests/filter_tests/test_truncatechars_html.py (hunk 11 above)
Comment (ngnpope) [reply tests/template_tests/filter_tests/test_truncatechars_html.py:44]: Ok, so this is definitely a regression we're introducing here that we should fix.

I added the following test to `tests/template_tests/filter_tests/test_truncatechars.py` which passes fine:

```python
    @setup({"truncatechars04": "{{ a|truncatechars:3 }}"})
    def test_truncatechars01(self):
        output = self.engine.render_to_string(
            "truncatechars04", {"a": "abc"}
        )
        self.assertEqual(output, "abc")
```

I think we should add this as it covers the exact length case which is missing there.

The HTML chars case is regressing. The problem is on this line:

https://github.com/django/django/pull/16421/files#diff-a6acaa0a744b3a7c841d9ab3ccbc1765f5749e3782fced61edced456c227b43fR188

We are passing in `length=truncate_len` when we should be passing in `length=length`, as for the words case. When you look at the text chars case we are passing through both `length` and `truncate_len`.

I suspect that we will want to hoist the following into a global helper:

https://github.com/django/django/pull/16421/files#diff-a6acaa0a744b3a7c841d9ab3ccbc1765f5749e3782fced61edced456c227b43fR180-R186

Which we can then call independently from the appropriate place in the text branch or the HTML branch (inside of `TruncateCharsHTMLParser`).

**Review by ngnpope — CHANGES_REQUESTED** [review]
Thanks David. Definitely feels like we're getting closer.

**Comment by nessita** [comment]
@smithdc1 Hey! Is this ready for re-review? If yes perhaps we should unset the needs improvement flag in the ticket.

File: docs/releases/5.0.txt [hunk 19]
```diff
@@ -576,6 +576,10 @@ Miscellaneous
 * The minimum supported version of SQLite is increased from 3.21.0 to 3.27.0.
 
 * Support for ``cx_Oracle`` < 8.3 is removed.
+* ``django.utils.text.Truncator`` used by :tfilter:`truncatechars_html` and 
+  :tfilter:`truncatewords_html` now uses an HTML parser which is more 
+  robust. HTML entities are now decoded and there may be other small
```
Comment (nessita) [inline docs/releases/5.0.txt:581]: I'm not sure what this sentence means, what would be a decoded HTML entity? :thinking:

**Review by nessita — COMMENTED** [review]
Initial comments, looks good!

**Comment by felixxm** [comment]
> @smithdc1 Hey! Is this ready for re-review? If yes perhaps we should unset the needs improvement flag in the ticket.

There are unresolved comments, e.g. https://github.com/django/django/pull/16421#discussion_r1263859265 or https://github.com/django/django/pull/16421#discussion_r1263439549.

**Comment by nessita** [comment]
> There are unresolved comments, e.g. [#16421 (comment)](https://github.com/django/django/pull/16421#discussion_r1263859265) or [#16421 (comment)](https://github.com/django/django/pull/16421#discussion_r1263439549).

I believe the first one is already addressed in https://github.com/django/django/pull/16421/commits/4f1791c26ad14954e42def6c5dd71d362c94c4d5

**Comment by nessita** [comment]
@smithdc1 Hi! Happy New Year :fireworks: 

Would you have time to keep working on this? If not I might try to push it to the finish line. Let me know! Thanks.

**Comment by smithdc1** [comment]
Hi @nessita -- happy new year. Hope you had a good break! 🎆

Yes. Let me update this one. I'll have to remind myself of where I got to!

File: tests/utils_tests/test_text.py (hunk 3 above)
Comment (smithdc1) [reply tests/utils_tests/test_text.py:98]: These tests have since been re-written in 17b51094d778b421bb2b3aae0c270894b050455d

File: docs/releases/5.0.txt (hunk 19 above)
Comment (smithdc1) [reply docs/releases/5.0.txt:581]: I think I'm just trying to say there may be small differences in the output compared to the previous version. 

The release note is now:

```
* ``django.utils.text.Truncator`` used by :tfilter:`truncatechars_html` and
  :tfilter:`truncatewords_html` now uses a :py:class:`html.parser.HTMLParser` 
  in its underlying implementation. This change results in a more robust and 
  faster operation, but there may be small differences in the output. 

```

File: tests/utils_tests/test_text.py (hunk 8 above)
Comment (smithdc1) [reply tests/utils_tests/test_text.py]: Can I check my understanding of this? Are you saying to squash this PR to a single commit? (Happy to do so, just don't want to split it out again later...)

Comment (nessita) [reply tests/utils_tests/test_text.py]: @felixxm could you please clarify what you mean? David and I understand from your comment that the two commits would be better off squashed into one, is that your ask?

Comment (felixxm) [reply tests/utils_tests/test_text.py]: > @felixxm could you please clarify what you mean? David and I understand from your comment that the two commits would be better off squashed into one, is that your ask?

No, I'm talking only about the following assertion :
```
        self.assertEqual(
            "…",
            truncator.chars(0, html=True),
        )
```
It should not be a part of the first commit, it's a regression test that should be added as:
```
        self.assertEqual(
            "",
            truncator.chars(0, html=True),
        )
```
in the second commit (with the main change).

Comment (smithdc1) [reply tests/utils_tests/test_text.py]: Thanks for explaining. I've made this change.

File: django/utils/text.py [hunk 20]
```diff
@@ -104,21 +207,20 @@ def chars(self, num, truncate=None, html=False):
         """
         self._setup()
         length = int(num)
+        if length <= 0:
+            return ""
```
Comment (felixxm) [inline django/utils/text.py:211]: I moved this bugfix to a separate PR, #17832.

**Comment by felixxm** [comment]
@smithdc1 Thanks for all your efforts :+1:

**Comment by ngnpope** [comment]
Yes, thanks @smithdc1. This is a solid improvement. Also thanks for enduring my tweaks/reviews!

**Comment by pauloxnet** [comment]
Great improvement @smithdc1 👏🏻
